slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -646,6 +668,29 @@ void SLAMServiceImpl::RunSLAM() {
                 latest_map_filename.find(viam::io::filename_prefix) +
                     viam::io::filename_prefix.length(),
                 latest_map_filename.find(".pbstream")));
+
+        // If using the LOCALIZING action mode, cache a copy of the map before
+        // beginning to process data. If cartographer fails to do this,
+        // terminate the program
+        if (action_mode == ActionMode::LOCALIZING) {
+            try {
+                GetLatestSampledPointCloudMapString(
+                    localization_pointcloud_map);
+            } catch (std::exception &e) {
+                LOG(ERROR) << "Stopping Cartographer: error encoding localized "
+                              "pointcloud map: "
+                           << e.what();
+                std::terminate();
+            }
+
+            if (localization_pointcloud_map.empty()) {
+                LOG(ERROR) << "Stopping Cartographer: error encoding localized "
+                              "pointcloud map: no map points";
+                std::terminate();
+            }
+
+            localization_map_ready = true;
+        }
```

Comment:
[q] Could we run this for both updating & localizing mode and save the map in `latest_pointcloud_map` instead? That way, we'd also have a map for "updating" at startup - and when no data has been added yet (e.g., maybe a user wants to view the map they load without running cartographer with new data?)

Reply:
There's no reason to run at startup in updating mode. They could just ask for a map and receive the same result. In addition, there is no way for updating mode to know whether to use the backup map vs just asking for a new map and to make it smart enough for the scenario you suggest is out of scope for this ticket.